File: src/files/TemplateFileBuilder.java
/*
 * Copyright (c) 2015. This code is a LogosProg property. All Rights Reserved.
 */

package files;

import system.ConsoleMessage;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Created by forando on 14.06.15.<br>
 * This class constructs external system files <b>from default
 * template files</b>.
 */
public abstract class TemplateFileBuilder<T> extends SystemFileManager implements ObjectFromFileBuilder<T, InputStream> {

    /**
     * The object of predefined type that we construct from external file
     */
    private T mainObj = null;


    /**
     *
     * @param fileName A file name that will be used to construct {@link T} object from
     * @param rootDir Application root directory
     * @param subDir Optional. A subdirectory name the file will be located
     * @throws IOException If either <b>fileName</b> or <b>rootDir</b> is null
     */
    public TemplateFileBuilder(String fileName, String rootDir, String subDir) throws IOException {
        super(fileName, rootDir, subDir);
    }

    @Override
    public T build(InputStream element) throws IOException {
        if (fileExists()){
            mainObj = getObjectFromExternalFile();
        }else{
            boolean created = generateDefaultFile(element);
            if (created){
                mainObj = getObjectFromExternalFile();
                if (listener != null) listener.onFileGenerated();
            }else {
                ConsoleMessage.printErrorMessage("An error occurred during " + getFileName() + " file creation.");
            }
        }
        return mainObj;
    }

    /**
     * Implement this method to construct for user a {@link T} object from external file
     * @return a {@link T} object
     * @throws IOException If {@link T} object construction failed.
     */
    protected abstract T getObjectFromExternalFile() throws IOException;

    /**
     * Implement this method to return for user {@link #mainObj} <b>if it's
     * been already created</b>
     * @return {@link #mainObj}
     */
    protected T getMainObject(){
        return mainObj;
    }

    /**
     * This method must be implemented in order to save a {@link T}
     * object back to external file.
     * @throws IOException If saving to external file failed.
     */
    protected abstract void setMainObject()throws IOException;

    /**
     * This method copies default file from internal app folders to specific
     * project folder.
     * @return True - if operation is successful.
     */
    protected boolean generateDefaultFile(InputStream is){

        boolean result = false;

        OutputStream os = null;
        String path;

        if (null != subDir) {
            path = getFilePath();
        }else{
            path = getFilePath();
        }

        if (!createEmptyFile()) return false;

        try {

            //is = SystemFileManager.class.getResourceAsStream(fileName);
            os = new FileOutputStream(path);
            byte[] buffer = new byte[1024];
            int length;
            while ((length = is.read(buffer)) > 0) {
                os.write(buffer, 0, length);
            }
            result = true;
        }catch(IOException ex){
            ex.printStackTrace();
        }finally {
            if (is != null) {
                try {
                    is.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            if (os != null) {
                try {
                    os.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return result;
    }

    TemplateFileBuilderListener listener;

    public void addTemplateFileBuilderListener(TemplateFileBuilderListener listener){
        this.listener = listener;
    }


    public interface TemplateFileBuilderListener{
        /**
         * Notifies when the desired file has been just generated from the default template file.
         */
        void onFileGenerated();
    }
}

File: src/main/kotlin/com/logosprog/mainutils/files/builders/TemplateFileBuilder.kt
/*
 * Copyright (c) 2016. This code is a LogosProg property. All Rights Reserved.
 */

package com.logosprog.mainutils.files.builders

import com.logosprog.mainutils.files.managers.SystemFileManager
import com.logosprog.mainutils.system.ConsoleMessage1
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

/**
 * Constructs external system files **from default
 * template files**.
 */
abstract class TemplateFileBuilder<T>
/**

 * @param fileName A file name that will be used to construct [T] object from
 * *
 * @param rootDir Application root directory
 * *
 * @param subDir Optional. A subdirectory name the file will be located
 * *
 * @throws IOException If either **fileName** or **rootDir** is null
 */
@Throws(IOException::class)
constructor(fileName: String, rootDir: String, subDir: String) : SystemFileManager(fileName, rootDir, subDir), ObjectFromFileBuilder<T, InputStream> {

    private val TAG: String

    /**
     * The object of predefined type that we construct from external file
     */
    /**
     * Implement this method to return for user [.mainObj] **if it's
     * been already created**
     * @return [.mainObj]
     */
    protected var mainObject: T? = null
        private set


    init {
        TAG = this.javaClass.simpleName
    }

    @Throws(IOException::class)
    override fun build(element: InputStream): T {
        if (fileExists()) {
            mainObject = objectFromExternalFile
        } else {
            val created = generateDefaultFile(element)
            if (created) {
                mainObject = objectFromExternalFile
                if (listener != null) listener!!.onFileGenerated()
            } else {
                ConsoleMessage1.printErrorMessage("$TAG.build: An error occurred during $fileName file creation.")
            }
        }
        return mainObject as T
    }

    /**
     * Implement this method to construct for user a [T] object from external file
     * @return a [T] object
     * *
     * @throws IOException If [T] object construction failed.
     */
    protected abstract val objectFromExternalFile: T

    /**
     * This method must be implemented in order to save a [T]
     * object back to external file.
     * @throws IOException If saving to external file failed.
     */
    @Throws(IOException::class)
    protected abstract fun setMainObject()


    /**
     * This method copies default file from internal app folders to specific
     * project folder.
     * @param inputStream Input stream to generate the file from
     * *
     * @return True - if operation is successful.
     */
    protected fun generateDefaultFile(inputStream: InputStream): Boolean {
        val path = getFilePath()

        if (!createEmptyFile()){
            inputStream.close()
            return false
        }

        val outputStream: OutputStream? = FileOutputStream(path)
        val buffer = ByteArray(1024)
        var length: Int = 0

        while (inputStream.read(buffer).let{
            length = it
            it > 0
        }){
            outputStream?.write(buffer, 0, length)
        }

        return true
    }

    internal var listener: TemplateFileBuilderListener? = null

    fun addTemplateFileBuilderListener(listener: TemplateFileBuilderListener) {
        this.listener = listener
    }


    interface TemplateFileBuilderListener {
        /**
         * Notifies when the desired file has been just generated from the default template file.
         */
        fun onFileGenerated()
    }
}
